dataflow/snippets/src/main/java/com/example/dataflow/KafkaRead.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.example.dataflow;
+
+// [START dataflow_kafka_read]
+import com.google.common.collect.ImmutableMap;
+import java.io.UnsupportedEncodingException;
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.io.TextIO;
+import org.apache.beam.sdk.managed.Managed;
+import org.apache.beam.sdk.options.Description;
+import org.apache.beam.sdk.options.PipelineOptionsFactory;
+import org.apache.beam.sdk.options.StreamingOptions;
+import org.apache.beam.sdk.transforms.MapElements;
+import org.apache.beam.sdk.values.TypeDescriptors;
+
+public class KafkaRead {
+
+  public interface Options extends StreamingOptions {
+    @Description("The Kafka bootstrap server. Example: localhost:9092")
+    String getBootstrapServer();
+
+    void setBootstrapServer(String value);
+
+    @Description("The Kafka topic to read from.")
+    String getTopic();
+
+    void setTopic(String value);
+
+    @Description("Path to write the output file")
+    String getOutputPath();
+
+    void setOutputPath(String value);
+  }
+
+  public static void main(String[] args) {
+    // Parse the pipeline options passed into the application. Example:
+    //   --bootstrap_servers=$BOOTSTRAP_SERVERS --topic=$KAFKA_TOPIC --outputPath=$OUTPUT_FILE
+    // For more information, see https://beam.apache.org/documentation/programming-guide/#configuring-pipeline-options
+    var options = PipelineOptionsFactory.fromArgs(args).withValidation().as(Options.class);
+    options.setStreaming(true);
+
+    var pipeline = Pipeline.create(options);
+    ImmutableMap<String, Object> config = ImmutableMap.<String, Object>builder()
+        .put("bootstrap_servers", options.getBootstrapServer())
+        .put("topic", options.getTopic())
+        .put("data_format", "RAW")
+        .put("max_read_time_seconds", 15)
+        .put("auto_offset_reset_config", "earliest")
+        .build();
+
+    // Build the pipeline.
+    pipeline
+        // Read messages from Kafka.
+        .apply(Managed.read(Managed.KAFKA).withConfig(config)).getSinglePCollection()
```

Comment:
nit: I am not SME in using Kafka with Apache Beam (or Dataflow). What is a reason for using configuration two times: one as Options and another in the first `apply()` call?

Reply:
According to your clarifications, the use of `Options` is done for convenience so you can avoid coding a complex parsing logic. Given explanations provided in the [previous comment](https://github.com/GoogleCloudPlatform/java-docs-samples/pull/9424#discussion_r1714640234), please consider separating the parsing logic that effectively can be left at user's discretion, from the recommended method of reading from Kafka that requires the "property bag of configuration parameters". I would recommend to pass the parameters as arguments to the method that will be shown in the code snippet. In this form, the parsing of the arguments that you choose to use will not be part of the snippet and users will be able to implement it at their discretion.
This refactoring will also simplify testing the method since parameters can be explicitly set up in the testing method.

Reply:
I need to pass a PipelineOptions regardless, because that tells Beam which runner to use, along with any Dataflow runtime options, streaming options, etc. For testing I'm using the DirectRunner, and Beam parses that via `Pipeline.create(options)`. 

The standard way to pass custom options to a Dataflow pipeline is by extending PipelineOptions, as shown. Splitting them out from the pipeline options isn't a best practice.